docs/rfcs/0006-top-sites-feature.md

```diff
@@ -0,0 +1,125 @@
+---
+layout: page
+title: Introduce a Top Sites Feature
+permalink: /rfc/0006-top-sites-feature
+---
+
+* Start date: 2020-07-30
+* RFC PR: [#7931](https://github.com/mozilla-mobile/android-components/pull/7931)
+
+## Summary
+
+Create a TopSitesFeature that abstracts out the co-ordination between the multiple data storages.
+
+## Motivation
+
+A large part of the top sites implementation is spread across A-C and Fenix. There is a lot of co-ordination between the History API and the current Top Sites API to manage the final state of what sites should be displayed that can exist as a single implementation.
+
+## Guide-level explanation
+
+To begin, we should rename the `TopSitesStorage`  to `PinnedSitesStorage` and create a `TopSitesFeature` that would orchestrate the different storages and present them to the app.
+
+We can then follow the architecture of existing features by using the presenter/interactor/view model, and introduce a `TopSitesStorage` which abstract out the complexity of the multiple data sources from `PinnedSitesStorage` and `PlacesHistoryStorage`.
+
+To allow the ability for adding a top site from different parts of the app (e.g. context menus), we introduce the `TopSitesUseCases` that acts on the storage directly. If the `TopSitesFeature` is started, it will then be notified by the `Observer` to update the UI.
+
+```kotlin
+
+/**
+ * Implemented by the application for displaying onto the UI.
+ */
+interface TopSitesView {
+  /**
+   * Update the UI.
+   */
+  fun displaySites(sites: List<TopSite>)
+
+  interface Listener {
+    /**
+     * Invoked by the UI for us to add to our storage.
+     */
+    fun onAddTopSite(topSite: TopSite)
+
+    /**
+     * Invoked by the UI for us to remove from our storage.
+     */
+    fun onRemoveTopSite(topSite: TopSite)
+  }
+}
+
+/**
+ * Abstraction layer above the multiple storages.
+ */
+interface TopSitesStorage {
+
+  /**
+   * Return a unified list of top sites based on the given number of sites desired.
+   * If `includeFrecent` is true, fill in any missing top sites with frecent top site results.
+   */
+  fun getTopSites(totalNumberOfSites: Int, includeFrecent: Boolean): List<TopSite>
+
+  interface Observer {
+    /**
+     * Invoked when changes are made to the storage.
+     */
+    fun onStorageUpdated()
+  }
+}
+
+// Already exists in browser-storage-sync.
+interface PlacesHistoryStorage {
+  fun getTopFrecentSites(num: Int)
+}
+
+class DefaultTopSitesStorage(
+  val pinnedSitesStorage: PinnedSitesStorage,
+  val historyStorage: HistoryStorage? = null
+) : TopSitesStorage {
+
+  /**
+   * Merge data sources here, return a single list of top sites.
+   */
+  override fun getTopSites(numFrecent): List<TopSite>
```

Comment:
`numFrecent` should just be `num`. Application has no way to know how many frecent sites there are, and neither does it care. What it cares is how many "top sites" it wants to display, and this storage is responsible for providing that number of top sites (some pinned, some frecent, some default).

Reply:
Oops, typo.

Reply:
TODO

- [ ] Change `numFrecent` to `limit` or `num`

Reply:
> TODO
> 
>     * [ ]  Change `numFrecent` to `limit` or `num`

It should match the API from the `TopSitesStorage` since it'll be an override: 

```suggestion
  override fun getTopSites(totalNumberOfSites: Int, includeFrecent: Boolean): List<TopSite>
```